### ISSUE TITLE
Explore Machine Learning Approach To Analysis Termination

### ISSUE THREAD

**Issue by notowen333** [privacy-tech-lab/privacy-pioneer#311]
Linked to #297 

We've already limited the set of requests to <100k characters for analysis. One idea is to look at machine learning techniques for deciding whether or not to analyze a request. 

We know that ~<10% of requests are going to yield a label. Maybe machine learning helps us cut corners to be more selective in our analysis.

## Steps
- Create a known set of requests that led to labels
- Train our model on this set
- Use this model to make the decision to analyze a request or not

## Background Exploration
- We'll need to find out how people do machine learning in javascript/browser extensions.
- We'd do well to get in contact with someone at the university who knows more than us about the topic.

## Stray Thoughts

- A different model for different request lengths?
- Labeled requests should probably come from something like top 100 alexa sites.
- Maybe the model can be trained on a subset of the request like the first 5000 characters or some other aspects of the request
-  Resource type as an additional training/labeling factor?

**Comment by SebastianZimmeck** [comment]
@notowen333 and @danielgoldelman, if we decide to go this route, it would be good if you can keep track of the sites that you are using to develop the machine learning approach (the validation set). The reason is that we do not want to use this validation set to test our approach. For the testing we need a separate set of unseen sites (the test set). Feel free to create a Google Doc or Sheet in our Google Drive to keep track of the already used sites (or you can use any other method that you prefer).

**Comment by notowen333** [comment]
If we were going to do this ourselves, we would probably want to look into [ml5](https://ml5js.org/) instead of [tensorflow.js](https://www.tensorflow.org/js) because it is a simpler library. 

However, as of now, it's not necessary to use machine learning to decide what to analyze because we are already performant enough.
